### PR TITLE
MAGN-4858 Adaptive components placed at origin when points are null

### PR DESCRIPTION
**Pull request by kiki-zhang**
The change includes a fix of MAGN-4858.
Root Cause: For creating a adaptive component, multiple RevitAPI are
called in one transaction. The first step is to create a original
component, and then call another RevitAPI to set its position. The issue
happens because an exception throws out in the process of setting
position due to the null point, which finally leads to the uncompleted
creation.
Solution: Using a subtransaction to roll back the creation if an
exception happens.

### CODE REVIEW

**Comment by sharadkjaiswal** [comment]
@Randy-Ma you might be interested in it.

**Comment by kiki-zhang** [comment]
@ikeough @Randy-Ma @DynamoDS/michelangelo Please help do the code review, thanks.

**Comment by ikeough** [comment]
@kiki-zhang Really excited to see PRs from the Revit team. I've added one comment that applies to all the places where we construct ACs.

Thanks!

**Comment by kiki-zhang** [comment]
Ian, thanks for your comments.
Regarding the performance concern for huge amounts of ACs creation, we have introduced an API change in Revit2016 to support the batch creation, it should be more efficient than placing AC one by one. If needed, we can consider to extend the batch creation in Dynamo.

**Comment by kronz** [comment]
Hi @kiki-zhang .  There is actually a task in the Youtrack backlog about this, not sure if you have access:
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-3427
I'm planning on talking to Alley about this soon.

**Comment by kiki-zhang** [comment]
@kronz I can access to this link, and add myself as watcher now. ACs batch creation is supported since Revit2016, I believe we cannot make it in DynamoRevit 2015 branch.

**Comment by kiki-zhang** [comment]
I am not familiar with Github and got involving several commits, sorry for that. I updated the fix according to @ikeough comments, trying to throw out more detail information about failure. The detail info we get from Revit as it has already handles intern errors and throw them out, we can show them directly.
With a null value, it throws with message such as "The value is null..". With a non-adaptive family instance, it throws such as "The Symbol famSymb is not an Adaptive Family Symbol..."...

File: src/Libraries/RevitNodes/Elements/AdaptiveComponent.cs [hunk 1]
```diff
@@ -88,24 +87,35 @@ private void InitAdaptiveComponent(Point[] pts, FamilySymbol fs)
...
+                    st.Start();
+                    var fam = AdaptiveComponentInstanceUtils.CreateAdaptiveComponentInstance(Element.Document, fs.InternalFamilySymbol);
+                    InternalSetFamilyInstance(fam);
+                    InternalSetPositions(pts.ToXyzs());
+                    st.Commit();
+                }
+                catch (Exception)
+                {
+                    st.RollBack();
+                    throw new ArgumentException("An adaptive component could not be found or created.");
```
Comment (Randy-Ma) [inline src/Libraries/RevitNodes/Elements/AdaptiveComponent.cs]: The messages need to be added to the resource file to support localization.

**Comment by kiki-zhang** [comment]
Thanks @Randy-Ma .
I updated the exception message in subsequent commit to display more detail error info, and now the new message includes two parts, one part is the original text "An adaptive component could not be created.", and the other part extracts from RevitAPI or System Exception.
We may have bilingual message if having the original text translated and leave others. Is it acceptable?

**Comment by ikeough** [comment]
@kiki-zhang Thanks for the updates, and sorry for the delayed reply. 

We should move the first part of the string to a resource file, so we don't have a dual language message. We will cover that in a separate fix.

The fix looks sufficient to cover my original requests. Please merge and move the associated YouTrack issue to "Testing". LGTM.

**Pull request closed without merging**